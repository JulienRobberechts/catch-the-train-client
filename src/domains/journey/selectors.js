export const selectCurrentJourney = (state) => {
  return state?.journey;
};
