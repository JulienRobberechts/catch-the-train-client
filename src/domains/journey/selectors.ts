import { RootState } from "../../redux-store/types";

export const selectCurrentJourney = (state: RootState) => {
  return state?.journey;
};
