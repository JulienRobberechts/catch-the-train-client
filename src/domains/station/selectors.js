export const selectUserConfiguration = (state) =>
  state?.station?.userConfiguration;

export const selectStationConfigurations = (state) =>
  state?.station?.stationConfigurations;
