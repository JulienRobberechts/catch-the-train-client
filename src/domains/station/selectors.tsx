import { RootState } from "../../redux-store/types";

export const selectUserConfiguration = (state: RootState) =>
  state?.station?.userConfiguration;

export const selectStationConfigurations = (state: RootState) =>
  state?.station?.stationConfigurations;
