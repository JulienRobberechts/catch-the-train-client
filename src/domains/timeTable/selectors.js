export const selectConfigIsValid = (state) => {
  return true;
  // Not sure it should be here!!
  // return !!state?.timeTable.route?.station?.code;
};

export const selectRoute = (state) => {
  return state?.timeTable?.data;
};

export const selectTimeTableContext = (state) => {
  return state?.timeTable?.data?.context;
};

export const selectTrainDeparture = ({
  network,
  line,
  station,
  train: trainCode,
}) => (state) => {
  console.log("selectTrainDeparture");

  // todo:
  // check network, line, station are still ok.

  // search in trains
  const departures = state?.timeTable?.data?.departures;

  if (!departures) {
    return null;
  }

  // TODO ...

  const currentIndex = Math.max(
    0,
    departures.findIndex((departure) => departure.trainCode === trainCode)
  );

  return departures[currentIndex];

  // if found:
  // if not found: return closest to the time
  // return `MY TRAIN DEPARTURE trainCode=${trainCode} currentIndex=${currentIndex}`;
};
