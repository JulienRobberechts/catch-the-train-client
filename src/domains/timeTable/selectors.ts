import { getMatchingDeparture } from "./pure";
import { RootState } from "../../redux-store/types";

export const selectRequestStatus = (state: RootState) => {
  return {
    loading: !!state?.timeTable?.loading,
    error: state?.timeTable?.error,
    hasData: !!state?.timeTable?.data,
  };
};

export const selectAllDepartures = (state: RootState) => {
  return state?.timeTable?.data?.departures ?? [];
};

export const selectDepartureByTrainCode = (trainCode: string) => (
  state: RootState
) => {
  const allDepartures = state?.timeTable?.data?.departures;

  if (!allDepartures) {
    return null;
  }
  return getMatchingDeparture(allDepartures, trainCode);
};
