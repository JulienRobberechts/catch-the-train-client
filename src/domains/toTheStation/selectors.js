import moment from "moment";
import { getDelay, getDelayStatus } from "./pure";

// to move
import { selectAllDepartures } from "../timeTable/selectors";

export const selectNow = (state) => state?.toTheStation?.currentTime;

// To split into smaller piece
export const selectToTheStation = (state) => state?.toTheStation;

// TO MOVE ??
export const selectEnhancedToTheStation = (state) => {
  // const { timeTable } = state; // MASHUP !!!!!!

  const toTheStation = selectToTheStation(state);
  const departures = selectAllDepartures(state);

  if (!departures || !toTheStation) return null;

  const trainCode = toTheStation?.train?.trainCode;

  if (!trainCode) {
    return null;
  }

  const departureIndex = Math.max(
    departures.findIndex((departure) => departure.trainCode === trainCode),
    0
  );

  const departure = departures[departureIndex];

  const {
    userConfiguration: { onTimeMarginDelaySeconds },
    stationConfiguration: { travelDurationSeconds, waitingDelaySeconds },
  } = toTheStation;

  const travelDuration = moment.duration({
    seconds: travelDurationSeconds,
  });
  const waitingDuration = moment.duration({
    seconds: waitingDelaySeconds,
  });
  const targetTime = moment.parseZone(departure.departureTime);

  const nowTime = toTheStation.currentTime
    ? moment.parseZone(toTheStation.currentTime)
    : null;

  const { targetDuration, delayDuration } = getDelay({
    nowTime,
    targetTime,
    travelDuration,
    waitingDuration,
  });
  const delayStatus = getDelayStatus(delayDuration, onTimeMarginDelaySeconds);

  return {
    route: departure.route,
    nowTime,
    targetDuration,
    targetTime,
    travelDuration,
    waitingDuration,
    delayDuration,
    delayStatus,
  };
};
