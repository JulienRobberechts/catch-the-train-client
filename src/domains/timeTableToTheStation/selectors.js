import moment from "moment";
import { getDelay } from "../toTheStation/pure";

// those selectors are a mashup of timeTable and toTheStation
// in order to provide a ready to use selectors for the timeLine component.

// Be careful with change of references!!!
export const selectEnhancedTimeTable = state => {
  if (!state.timeTable.route || state.toTheStation.noData) return null;

  const { timeTable } = state;
  const nowTime = moment.parseZone(state.toTheStation.currentTime);

  const {
    configuration: { waitingDelaySeconds },
    station: { travelDurationSeconds, onTimeMarginDelaySeconds }
  } = state.toTheStation;

  const trains = state.timeTable.route.trains.map((departure, index) => {
    const departureTime = moment.parseZone(departure.departureTime);
    const departureDuration = moment.duration(departureTime.diff(nowTime));
    const targetTime = moment.parseZone(departure.departureTime);
    const travelDuration = moment.duration({
      seconds: travelDurationSeconds
    });
    const waitingDuration = moment.duration({
      seconds: waitingDelaySeconds
    });

    const { targetDuration, delayDuration, delayStatus } = getDelay({
      nowTime,
      targetTime,
      travelDuration,
      waitingDuration,
      onTimeMarginDelaySeconds
    });

    return {
      index,
      departureTime,
      departureDuration,
      targetDuration,
      delayDuration,
      delayStatus
    };
  });

  return { route: timeTable.route, trains };
};
