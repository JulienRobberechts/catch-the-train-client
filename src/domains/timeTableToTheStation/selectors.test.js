import timeTableReducer, {
  initialState as timeTableInitialState,
} from "../timeTable/slice";

import toTheStationReducer, {
  initialState as toTheStationInitialState,
} from "../toTheStation/slice";

import { selectEnhancedTimeTable } from "./selectors";

describe("slice timeTable", () => {
  describe("initial state", () => {
    const rootState = {
      timeTable: timeTableInitialState,
      toTheStation: toTheStationInitialState,
    };
    it("'selectEnhancedTimeTable' should return falsy", () => {
      expect(selectEnhancedTimeTable(rootState)).toBeFalsy();
    });
  });
});
