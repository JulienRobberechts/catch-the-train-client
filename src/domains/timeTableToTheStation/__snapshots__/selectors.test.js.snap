// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`slice timeTable after 'mockTimeTable' action, state 'selectEnhancedTimeTable' should return data 1`] = `
Object {
  "route": Object {
    "direction": "Paris",
    "directionsAliases": Array [
      "Paris",
      "Châtelet–Les Halles",
    ],
    "station": Object {
      "code": "SGL",
      "name": "Saint-Germain-en-Laye",
    },
    "trains": Array [
      Object {
        "departureTime": "2020-03-10T09:24:00+01:00",
        "platform": "2",
        "trainCode": "0924",
      },
      Object {
        "departureTime": "2020-03-10T09:32:00+01:00",
        "platform": "4",
        "trainCode": "0932",
      },
      Object {
        "departureTime": "2020-03-10T09:43:00+01:00",
        "platform": "2",
        "trainCode": "0943",
      },
      Object {
        "departureTime": "2020-03-10T09:55:00+01:00",
        "platform": "4",
        "trainCode": "0955",
      },
      Object {
        "departureTime": "2020-03-10T10:05:00+01:00",
        "platform": "4",
        "trainCode": "1005",
      },
      Object {
        "departureTime": "2020-03-10T10:21:00+01:00",
        "platform": "2",
        "trainCode": "1021",
      },
      Object {
        "departureTime": "2020-03-11T12:12:00+01:00",
        "platform": "2",
        "trainCode": "1212",
      },
    ],
  },
  "trains": Array [
    Object {
      "delayDuration": "-PT215H29M29.621S",
      "delayStatus": "late",
      "departureDuration": "-PT215H17M24.621S",
      "departureTime": "2020-03-10T08:24:00.000Z",
      "index": 0,
      "targetDuration": "-PT215H17M24.621S",
      "trainCode": "0924",
    },
    Object {
      "delayDuration": "-PT215H21M29.621S",
      "delayStatus": "late",
      "departureDuration": "-PT215H9M24.621S",
      "departureTime": "2020-03-10T08:32:00.000Z",
      "index": 1,
      "targetDuration": "-PT215H9M24.621S",
      "trainCode": "0932",
    },
    Object {
      "delayDuration": "-PT215H10M29.621S",
      "delayStatus": "late",
      "departureDuration": "-PT214H58M24.621S",
      "departureTime": "2020-03-10T08:43:00.000Z",
      "index": 2,
      "targetDuration": "-PT214H58M24.621S",
      "trainCode": "0943",
    },
    Object {
      "delayDuration": "-PT214H58M29.621S",
      "delayStatus": "late",
      "departureDuration": "-PT214H46M24.621S",
      "departureTime": "2020-03-10T08:55:00.000Z",
      "index": 3,
      "targetDuration": "-PT214H46M24.621S",
      "trainCode": "0955",
    },
    Object {
      "delayDuration": "-PT214H48M29.621S",
      "delayStatus": "late",
      "departureDuration": "-PT214H36M24.621S",
      "departureTime": "2020-03-10T09:05:00.000Z",
      "index": 4,
      "targetDuration": "-PT214H36M24.621S",
      "trainCode": "1005",
    },
    Object {
      "delayDuration": "-PT214H32M29.621S",
      "delayStatus": "late",
      "departureDuration": "-PT214H20M24.621S",
      "departureTime": "2020-03-10T09:21:00.000Z",
      "index": 5,
      "targetDuration": "-PT214H20M24.621S",
      "trainCode": "1021",
    },
    Object {
      "delayDuration": "-PT188H41M29.621S",
      "delayStatus": "late",
      "departureDuration": "-PT188H29M24.621S",
      "departureTime": "2020-03-11T11:12:00.000Z",
      "index": 6,
      "targetDuration": "-PT188H29M24.621S",
      "trainCode": "1212",
    },
  ],
}
`;
